### PR TITLE
(Towards #342) tighten-up ACCRoutineTrans.validate() and test

### PR DESCRIPTION
**Pull request by arporter**

### CODE REVIEW

**Comment by arporter** [comment]
Unfortunately, the problem is that the imported symbol is just a `Symbol` while the one obtained by finding its definition is a `DataSymbol`. Can we morph a `Symbol` into a `DataSymbol` without doing something really dirty?

**Comment by arporter** [comment]
I think the best solution is to do the swap properly and walk through the tree and update every Reference. I could make this a new method of `Schedule` and have it update the SymbolTable too or I could extend the existing `SymbolTable.swap` method. I've done the latter at the moment but am not too comfortable with a method on the SymbolTable updating the nodes it is associated with. What do @sergisiso, @rupertford and @hiker think?

**Comment by sergisiso** [comment]
> Can we morph a Symbol into a DataSymbol without doing something really dirty?

Isn't this what the Symbol.specialise() does? The original intent of specialise was also to avoid doing the replacement of all references.

**Comment by arporter** [comment]
> > Can we morph a Symbol into a DataSymbol without doing something really dirty?
> 
> Isn't this what the Symbol.specialise() does? The original intent of specialise was also to avoid doing the replacement of all references.

Ah! Yes. Thanks Sergi. I'll see if I can use that instead as that's much nicer.

**Comment by sergisiso** [comment]
This may solve it for the resolve_deferred (and is probably the right way because it will fail if the import subclass is in contradiction of some of the local inferred attributes which can't happen).

However, the swap (and also a safe remove method) may still need to be fixed, but currently its hard to keep track of all references (as they may not be on the tree, e.g. init_values, precision symbols, ...) and we don't even guarantee that we don't reuse symbols in different trees/st (maybe we should?). One solution would be to implement a "observer pattern" for symbols, so they know all their uses. It could also ease some analysis and transformations.

**Comment by arporter** [comment]
The fparser2 frontend was already identifying that symbols used as kind parameters must be of integer type. According to the Fortran standard, they must also be compile-time constants so I've extended the frontend slightly to give them that property too. I've then extended the validation check in ACCRoutineTrans so that it permits imports of compile-time constants.

**Comment by arporter** [comment]
Integration tests were fine. CI is fine but is not reporting coverage.
Ready for a first look. Probably (another) one for @sergisiso as it changes details regarding Symbols. In particular, we now permit a Symbol with an ImportInterface or UnresolvedInterface to be a constant, even if we don't know its value.
With the changes here, it's possible to process all of gravity wave, add OpenACC and compile and run without having to script any exclusions.

File: src/psyclone/transformations.py
Comment (sergisiso) [inline src/psyclone/transformations.py:2488]: Maybe if there is any Codeblock at all, regardless of the symbols used, it should not validate as we can not guarantee that is going to work on the GPU.

Which mean that maybe we just need to add CodeBlock in the class exclude_node_type instead of this check?

File: src/psyclone/transformations.py
Comment (sergisiso) [inline src/psyclone/transformations.py:2465]: This is just a question/comment, not for this PR, but some transformations do not resolve deferred and instead produce validation errors, and then the script writer can resolve before calling the transformation (e.g. in nemo); another couple of places do the resolve when needed (like here), and in some cases it would be useful to have them earlier than the script, like during the reader (e.g. where cannonicalisation).

We should rethink/clarify what is the best practice here? And maybe even resolve from all containers that are listed on the "-I" files from the very beginning, so we always start with all information we have.

File: src/psyclone/tests/psyir/transformations/kernel_transformation_test.py
Comment (sergisiso) [inline src/psyclone/tests/psyir/transformations/kernel_transformation_test.py:380]: I would say that the outer parenthesis in this error message are not needed if we already have the single quotes.

**Review by sergisiso — CHANGES_REQUESTED** [review]
@arporter This is a good improvement for the ACCRoutine and I also agree with the change of specializing imports with information like is_constant and initial_value*. There are some places that our usage of the symbol table maybe could be improved and I mentioned in the inline comments.

File: src/psyclone/domain/lfric/kernel_interface.py
Comment (arporter) [reply src/psyclone/domain/lfric/kernel_interface.py:374]: I agree, it would be nicer. However, I'm not sure I can do that and still use the `LFRicTypes("NumberOfCellsDataSymbol")`. Perhaps the LFRic-specific symbol table should override `find_or_create` so that it can accept an `lfric_type` argument or similar?

Comment (arporter) [reply src/psyclone/domain/lfric/kernel_interface.py:374]: Actually, I realised that it's quite possible to use `find_or_create` so I've done that.

Comment (arporter) [reply src/psyclone/domain/lfric/kernel_interface.py:374]: Well, that proved complicated. Essentially, we *need* a new symbol table because we need to create symbols with the correct (argument) interfaces. I suspect ArgOrdering gets away with using the symbol table associated with the invoke Schedule because we aren't using interface information there (but we should). So, for now, I do need the `__init__` but I've cut it down and commented it.

File: src/psyclone/psyad/tl2ad.py
Comment (arporter) [reply src/psyclone/psyad/tl2ad.py:295]: The `symbol` here is a precision symbol and this code is really checking that we have a declaration for it because if we don't know where it has come from we can't safely generate the adjoint code. Having written that I've realised there's a nicer way of checking for this condition so I've changed the code. Hopefully it's better now.

File: src/psyclone/psyir/frontend/fparser2.py
Comment (arporter) [reply src/psyclone/psyir/frontend/fparser2.py]: I tried to use specialise in the next two branches but it all began to unravel - `specialise` complains if given a symbol of the same class as the one being specialised and I also hit problems because a bare Symbol doesn't have a datatype attribute. All in all, it was simpler to keep the special case for the `DataSymbol` which just modifies its properties.

File: src/psyclone/psyir/symbols/datasymbol.py
Comment (arporter) [reply src/psyclone/psyir/symbols/datasymbol.py]: Yes, that's better thanks. I had to allow for Symbols of UnknownType too as they can have an initialisation expression within them.

File: src/psyclone/psyir/symbols/datasymbol.py
Comment (arporter) [reply src/psyclone/psyir/symbols/datasymbol.py:176]: Done.

File: src/psyclone/psyir/symbols/symbol.py
Comment (arporter) [reply src/psyclone/psyir/symbols/symbol.py:245]: Yes, I think you're right.

File: src/psyclone/psyir/symbols/symbol.py
Comment (arporter) [reply src/psyclone/psyir/symbols/symbol.py:251]: Oh yes, not sure what happened there. Gone now.

File: src/psyclone/tests/psyir/symbols/symbol_test.py
Comment (arporter) [reply src/psyclone/tests/psyir/symbols/symbol_test.py:367]: Good idea, done.

File: src/psyclone/transformations.py
Comment (arporter) [reply src/psyclone/transformations.py:2521]: Oh yes :-( Added TODO.

File: src/psyclone/transformations.py
Comment (arporter) [reply src/psyclone/transformations.py:2529]: Yes, I'm beginning to think that would be best. Feels like a very similar conversation to the one in #2296.

File: src/psyclone/transformations.py
Comment (arporter) [reply src/psyclone/transformations.py:2552]: Unfortunately this transformation is not a `RegionTrans` so there's no exclude list. Therefore I've just altered the validate to reject CodeBlocks (unless the user forces it).

File: src/psyclone/tests/psyir/transformations/kernel_transformation_test.py
Comment (arporter) [reply src/psyclone/tests/psyir/transformations/kernel_transformation_test.py:380]: So would I.

File: src/psyclone/psyir/symbols/datasymbol.py
Comment (arporter) [inline src/psyclone/psyir/symbols/datasymbol.py:183]: @sergisiso: what would you prefer here - that I add an `is_unknown_type` property to Symbol or just add a pylint disable for the import?

**Comment by arporter** [comment]
Integration tests fail because one or more of the routines in `lib_fortran.f90` in NEMO contain a CodeBlock (because of #717). Happily, I can use the `force` option that I just implemented...

**Comment by sergisiso** [comment]
Note that CI for NEMO also breaks because we updated to 23.9, and merging https://github.com/stfc/PSyclone/pull/2385 will fix it. You may want to wait until  this is merged for running CI

**Comment by arporter** [comment]
Integration tests finally green :-)

**Review by sergisiso — CHANGES_REQUESTED** [review]
@arporter Another good improvement. All checks are green, docs build, codecov says we lost one line https://app.codecov.io/gh/stfc/PSyclone/pull/2366/indirect-changes could you check this? There are only  minor comments to address now. This can be merged when these are fixed.

File: examples/nemo/scripts/acc_kernels_trans.py
Comment (arporter) [reply examples/nemo/scripts/acc_kernels_trans.py:413]: Done.

File: src/psyclone/psyir/symbols/symbol.py
Comment (arporter) [reply src/psyclone/psyir/symbols/symbol.py:232]: Yes, done.

File: src/psyclone/tests/psyir/symbols/datasymbol_test.py
Comment (arporter) [reply src/psyclone/tests/psyir/symbols/datasymbol_test.py:69]: Oops. Updated.

**Comment by arporter** [comment]
That uncovered line is now unreachable due to the checks performed earlier so I've removed it (but also extended one of the tests too).

**Comment by arporter** [comment]
CI and integration tests permitting, this is ready for review again.

**Review by sergisiso — APPROVED** [review]
@arporter All comments have been addressed, coverage is now complete and tests still pass. There is a small docstring error which I fixed. This is ready to merge.

**Comment by sergisiso** [comment]
#2379 was ready first, so this is queued for merging after that one.

**Pull request merged**
